core-rust/core-api-server/src/core_api/conversions/lts.rs

```diff
@@ -365,13 +371,99 @@ pub fn to_api_lts_fungible_resource_balance_change(
+    let system_db_reader = SystemDatabaseReader::new(database);
+    let mut resultant_fungible_balances_by_account = HashMap::new();
+    substate_changes
+        .iter()
+        .filter_map(|(substate_reference, change_action)| {
+            let vault_id = &substate_reference.0;
+            if !vault_id.is_internal_fungible_vault() {
+                return None;
+            }
+
+            if substate_reference.1 != MAIN_BASE_PARTITION {
+                return None;
+            }
+
+            if substate_reference.2 != FungibleVaultField::Balance.into() {
+                return None;
+            }
+
+            let account_address = node_id_to_ancestor.get(vault_id).unwrap().parent.0;
+            let Ok(account_address) = GlobalAddress::try_from(account_address) else {
+                return None;
+            };
+
+            if !account_address.is_account() {
+                return None;
+            }
+
+            let vault_type_info = system_db_reader
+                .get_type_info(vault_id)
+                .expect("Vault missing TypeInfo substate");
+
+            let TypeInfoSubstate::Object(vault_type_info) = vault_type_info else {
+                panic!("TypeInfoSubstate of FungibleVault is not of type Object.");
+            };
+
+            assert!(
+                vault_type_info
+                    .blueprint_info
+                    .blueprint_id
+                    .eq(&BlueprintId::new(
+                        &RESOURCE_PACKAGE,
+                        FUNGIBLE_VAULT_BLUEPRINT,
+                    )),
+                "Vault's TypeInfo wrongly says this is not a fungible vault."
+            );
+
+            let resource_address =
+                ResourceAddress::new_or_panic(vault_type_info.get_outer_object().into());
+
+            let fungible_vault_balance_substate: FieldSubstate<VersionedFungibleVaultBalance> =
+                match change_action {
+                    ChangeAction::Create { new } => scrypto_decode(new).unwrap(),
+                    ChangeAction::Update { new, .. } => scrypto_decode(new).unwrap(),
+                    ChangeAction::Delete { .. } => {
+                        panic!("Invariant broken: vault substate is never deleted.")
+                    }
+                };
+
+            let resultant_balance = match fungible_vault_balance_substate.into_payload() {
+                VersionedFungibleVaultBalance::V1(liquid) => liquid.amount(),
+            };
+
+            Some((account_address, resource_address, resultant_balance))
+        })
+        .for_each(|(account_address, resource_address, resultant_balance)| {
+            resultant_fungible_balances_by_account
+                .entry(account_address)
+                .or_insert(Vec::new())
+                .push(models::LtsResultantFungibleBalance {
+                    resource_address: to_api_resource_address(context, &resource_address).unwrap(),
```

Comment:
```suggestion
                    resource_address: to_api_resource_address(context, &resource_address)?,
```

Reply:
I can change it into `.expect("some more clearer message")` but I would not pass that up as an "recoverable" error. I assume mapping error means that you (the client) passed up a wrong address, not that some random internal logic failed because invariants were broke. This clearly indicates something is wrong with the state, I think a panic is more suited.